platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/ResolutionHost.java

```diff
@@ -42,13 +50,71 @@ default DisplayName displayName(String type) {
     }
 
     /**
-     * Rethrows the provided failures. Does nothing if the list of failures is empty.
+     * Returns the problems service this host can use for reporting failures.
+     *
+     * @return the problems service
      */
-    default void rethrowFailure(String type, Collection<Throwable> failures) {
-        mapFailure(type, failures).ifPresent(e -> {
+    InternalProblems getProblems();
+
+    /**
+     * Rethrows the provided failures, doing nothing if the list of failures is empty.
+     * <p>
+     * If any of the failures (or their ancestor causes) are {@link ReportableAsProblem}, they will all be reported to the problems
+     * service available on this type via {@link #getProblems()}.
+     *
+     * @param resolutionType what was resolved, e.g. "dependencies", "artifacts", "files"
+     * @param failures the exceptions encountered during resolution
+     */
+    default void rethrowFailuresAndReportProblems(String resolutionType, Collection<Throwable> failures) {
+        reportProblems(failures);
+        consolidateFailures(resolutionType, failures).ifPresent(e -> {
             throw e;
         });
     }
 
-    Optional<? extends ResolveException> mapFailure(String type, Collection<Throwable> failures);
+    /**
+     * Consolidates the given failures into a single exception, if there are any.
+     *
+     * @param resolutionType what was resolved, e.g. "dependencies", "artifacts", "files"
+     * @param failures the exceptions encountered during resolution
+     * @return an {@link org.gradle.internal.exceptions.MultiCauseException} containing all the failures, or {@link Optional#empty()} if there are no failures
```

Comment:
🤔 I would go further to say this _always_ returns a `TypedResolveException`. We might even be able to guarantee that in the return type

Reply:
I think you're right.  I'll update the return type.